File: src/Functional/Haskell/Builder.hs
module Main where

import Data.List (intercalate)

data Builder = Builder
  { addTitle :: String -> [String] -> [String]
  , addSection :: String -> String -> [String] -> [String]
  , render :: [String] -> String
  }

textBuilder :: Builder
textBuilder = Builder
  { addTitle = \title parts -> parts ++ ["# " ++ title]
  , addSection = \heading body parts -> parts ++ ["## " ++ heading, body]
  , render = intercalate "\n"
  }

htmlBuilder :: Builder
htmlBuilder = Builder
  { addTitle = \title parts -> parts ++ ["<h1>" ++ title ++ "</h1>"]
  , addSection = \heading body parts -> parts ++ ["<h2>" ++ heading ++ "</h2>", "<p>" ++ body ++ "</p>"]
  , render = concat
  }

buildAvailabilityReport :: Builder -> String
buildAvailabilityReport builder =
  let withTitle = addTitle builder "Service status" []
      complete = addSection builder "Availability" "99.95%" withTitle
  in render builder complete

main :: IO ()
main = do
  putStrLn (buildAvailabilityReport textBuilder)
  putStrLn "---"
  putStrLn (buildAvailabilityReport htmlBuilder)

File: src/Functional/Clojure/builder.clj
(ns builder)

(defn text-builder []
  {:reset (fn [] [])
   :add-title (fn [title parts] (conj parts (str "# " title)))
   :add-section (fn [heading body parts] (conj parts (str "## " heading) body))
   :build (fn [parts] (clojure.string/join "\n" parts))})

(defn html-builder []
  {:reset (fn [] [])
   :add-title (fn [title parts] (conj parts (str "<h1>" title "</h1>")))
   :add-section (fn [heading body parts]
                  (conj parts (str "<h2>" heading "</h2>") (str "<p>" body "</p>")))
   :build (fn [parts] (apply str parts))})

(defn build-availability-report [builder]
  (let [parts ((:reset builder))
        parts ((:add-title builder) "Service status" parts)
        parts ((:add-section builder) "Availability" "99.95%" parts)]
    ((:build builder) parts)))

(println (build-availability-report (text-builder)))
(println "---")
(println (build-availability-report (html-builder)))
